jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -2337,6 +2337,23 @@ private void appendTypeParameters(StringBuilder nameBuilder) {
 			if (comma) nameBuilder.append(',');
 			nameBuilder.append(t);
 			comma = true;
+/*[IF Java12]*/
+			Type[] types = t.getBounds();
+			if (types.length == 1 && types[0].equals(Object.class)) {
+				// skip in case the only bound is java.lang.Object
+			} else {
+				boolean addExtends = true;
+				for (Type type : types) {
+					if (addExtends) {
+						nameBuilder.append(" extends "); //$NON-NLS-1$
+						addExtends = false;
+					} else {
+						nameBuilder.append(" & "); //$NON-NLS-1$
+					}
```

Comment:
I don't remember who initially suggested the pattern to me, but if you replace `boolean addExtends` with `String prefix` the loop is much simpler with in internal branching:
```
String prefix = " extends "; //$NON-NLS-1$
for (Type type : types) {
	nameBuilder.append(prefix).append(type.getTypeName());
	prefix = " & "; //$NON-NLS-1$
}
```

Reply:
Thanks for the suggestion. The code snippet proposed is going to create a string object for `" & "` regardless it is required to be appended or not (the usage without `&` are more common than those with it). This code carries more footprint than a primitive `boolean` variable which probably is not desirable.

Reply:
No, the string literal is in the constant pool in both your version and my proposal.

Reply:
Respectfully disagree. This PR does not create a string object `" & "` when there is only one `type` to be processed. The proposed code always creates such an object even when that string is not required to be appended.

Reply:
@JasonFengJ9 you're correct that if no one calls toGenericString() on a class with multiple type parameters the `" & "` will never be needed.  @keithc-ca is also correct that it's a one time cost - the first user of multiple type parameters pays the cost to create / intern it and everyone else benefits there after.

My preference is code patterns that are simpler to read & debug so I prefer Keith's as most code interacting with this api will pay for that second allocation at some point.

Reply:
Agreed. The string is going to be created/intern for future usage of `" & "`. Allowing this one time cost to favour the code pattern.
PR has been updated, please review.